doc/dev/ragged_iter_domain_design_doc.md

```diff
@@ -0,0 +1,914 @@
+
+### Terminology
+
+- **Ragged (or jagged) dimension**: A dimension with variable extent across batch elements
+- **Nested domains**: The individual IterDomains contained within a RaggedIterDomain, one per batch component
+- **Component**: One element of the batch (e.g., one sequence in a batch of sequences)
+- **Offset**: Starting position of each component in contiguous storage
+- **Extent**: Sum of all nested domain extents (total extent)
+
+### Expert Parallelism Use Case
+
+Expert parallelism in Mixture-of-Experts (MoE) models requires ragged dimensions. The key operations involve:
+
+1. **Token Dispatch**: Route input tokens to different experts based on routing decisions
+2. **Expert Distribution**: Distribute experts across multiple GPUs for parallel processing
+3. **Load Balancing**: Handle non-uniform token and expert distributions
+
+#### Single-Level Nesting Example
+
+After routing, tokens are grouped by expert with variable counts:
+
+```python
+# GPU 0 processes experts 0, 1, 2
+# Tokens per expert: [127, 0, 198]
+tokens_by_expert = nested_tensor([
+    tokens_for_expert_0,  # 127 tokens
+    tokens_for_expert_1,  # 0 tokens (expert not used)
+    tokens_for_expert_2   # 198 tokens
+])
+# Shape: [experts=3, tokens=[127,0,198], hidden=512]
+```
+
+This requires a single ragged dimension for tokens-per-expert.
+
+#### Multiple Independent Ragged Dimensions Example
+
+For load balancing, different GPUs may handle different numbers of experts based on their capacity:
+
+```python
+# GPU 0 handles 5 experts (experts 0-4, higher capacity)
+#   Expert 0: 127 tokens
+#   Expert 1: 0 tokens
+#   Expert 2: 198 tokens
+#   Expert 3: 64 tokens
+#   Expert 4: 89 tokens
+#   Total: 478 tokens on GPU 0
+
+# GPU 1 handles 3 experts (experts 5-7, lower capacity)
+#   Expert 5: 103 tokens
+#   Expert 6: 45 tokens
+#   Expert 7: 201 tokens
+#   Total: 349 tokens on GPU 1
+
+# Tensor representation:
+# Shape: [gpu=2, experts_per_gpu=[5,3], tokens_per_expert=[127,0,198,64,89,103,45,201], hidden=512]
+#
+# Breakdown by GPU:
+# - GPU 0: experts_per_gpu[0]=5, holds experts 0-4 with tokens [127,0,198,64,89]
+# - GPU 1: experts_per_gpu[1]=3, holds experts 5-7 with tokens [103,45,201]
+```
+
+This requires **two independent ragged dimensions**:
+1. **Expert distribution across GPUs**: `[5, 3]` - GPU 0 has 5 experts, GPU 1 has 3 experts
+2. **Token distribution per expert**: `[127, 0, 198, 64, 89, 103, 45, 201]` - each of the 8 total experts processes different numbers of tokens
+
+Both dimensions are ragged but independent (not nested within each other). The uneven expert distribution enables load balancing, with GPU 0 handling more experts (478 tokens total) and GPU 1 handling fewer experts (349 tokens total), resulting in somewhat balanced workload.
+
+---
+
+## 4. Goals and Non-Goals
+
+### Goals
+
+- Support ragged dimensions in fusion IR for expert parallelism
+- Support multiple independent ragged dimensions per tensor for load balancing scenarios
```

Comment:
We don't have to make up problems. Weights remain the dominating memory usage and should be split that as even as possible. Therefore, e is usually regularly split by d.

Reply:
See section 2.1.2 of https://arxiv.org/html/2412.19437v2